### PR TITLE
Check initializers of NT_CHECKED types for null terminators

### PR DESCRIPTION
**Pull request by Prabhuk**
Initializers of NT_CHECKED types must contain null terminators.
- If NT_CHECKED type arrays are part of other data types (e.g. structs), then the initializer of the parent type must be validated to make sure the NT_CHECKED contains a null terminator
- character arrays can be initialized with string literals. Must end with '\0' 
- Pointer to NT_CHECKED types when initialized, employ an implicit cast. Such cases are not handled as part of this task. 

Testing:
- Added checkedc tests for NT_CHECKED initializers
- Automated tests: DevTest Debug X64 Linux, DevTest Debug x64 Windows, DevTest Debug X86 Windows, DevTest Release Linux X64 LNT

Related to Issue: https://github.com/Microsoft/checkedc-clang/issues/397

### CODE REVIEW

File: lib/Sema/SemaDecl.cpp [hunk 1]
```diff
@@ -11394,47 +11496,47 @@ void Sema::ActOnUninitializedDecl(Decl *RealDecl) {
...
-          << RT->getDecl()->getTagKind() << Var;
-          break;
-        }
-        case Type::HasUncheckedPointer: {
-          Diag(Var->getLocation(), diag::err_initializer_expected_for_record_with_unchecked_pointer_in_checked_scope)
-          << RT->getDecl()->getTagKind() << Var;       
-          break;
-        }       
-        }
+
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: Did the white spacing on this get changed by your editor?    This code looks unchanged.  Please avoid unnecessary formatting changes, as it makes code harder to review.  It also makes merging the latest clang sources harder.

File: include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -9569,20 +9569,26 @@ def err_bounds_type_annotation_lost_checking : Error<
...
-  def err_initializer_expected_for_unchecked_pointer_in_checked_scope_with_bounds_expr : Error<
-    "unchecked pointer variable %0 in a checked scope with a bounds expression must have an initializer">;
-
-  def err_initializer_expected_for_array : Error<
-    "array variable %0 with elements containing checked pointers must have an initializer">;
-	
-  def err_initializer_expected_for_record_with_checked_value : Error<
+
+  def err_initializer_not_null_terminated_for_nt_checked : Error<
+      "null terminator expected in NT_CHECKED array initializer">;
```
Comment (dtarditi) [inline include/clang/Basic/DiagnosticSemaKinds.td]: Some of these diagnostics look unchanged.  Please avoid unnecessary white space changes.   Formatting changes should be done separately.

File: lib/Sema/SemaDecl.cpp [hunk 3]
```diff
@@ -11262,6 +11270,100 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit,
   CheckCompleteVariableDeclaration(VDecl);
 }
 
+/// Checked C: Validate that the NT_CHECKED type array initializers must be null terminated
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: Please add some comments explaining the arguments more: Init is an init expression for a declaration with VDeclType.

File: lib/Sema/SemaDecl.cpp [hunk 4]
```diff
@@ -11262,6 +11270,100 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit,
...
+      // Validating initializers of pointers require further analysis
+      return true;
+    }
+    case Type::ConstantArray: {
+      if (current->isNtCheckedArrayType()) {
+        const ConstantArrayType *CAT = Ctx.getAsConstantArrayType(VDeclType);
+        const auto DeclaredArraySize = CAT->getSize().getRawData();
+        InitListExpr *InitEx = dyn_cast<InitListExpr>(Init);
+        StringLiteral *InitializerString = nullptr;
+        if(InitEx && InitEx->getNumInits() == 1 &&
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: Please add a single line comment explaining what's going on here.  You are handling the case of a bracketed string { " … " }

File: lib/Sema/SemaDecl.cpp [hunk 5]
```diff
@@ -11262,6 +11270,100 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit,
...
+/// Checked C: Validate that the NT_CHECKED type array initializers must be null terminated
+bool Sema::ValidateNTCheckedType(ASTContext &Ctx, QualType VDeclType, Expr *Init) {
+  if(!Init) {
+    return true;
+  }
+  const Type *current = VDeclType.getTypePtr();
+  switch (current->getTypeClass()) {
+    case Type::Pointer: {
+      // Pointers to NT_CHECKED types can be initialized by any Expr option (e.g., calling a function)
+      // Validating initializers of pointers require further analysis
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: This comment is not quite accurate.    Epressions with type nt_array_ptr point to valid null-terminated values.   This is enforced static by type checking and checking of casts.  We don't need to check further here.

File: include/clang/Basic/DiagnosticSemaKinds.td [hunk 6]
```diff
@@ -9569,6 +9569,12 @@ def err_bounds_type_annotation_lost_checking : Error<
 
   def err_initializer_expected_for_ptr : Error<
     "automatic variable %0 with _Ptr type must have initializer">;
+
+  def err_initializer_not_null_terminated_for_nt_checked : Error<
+    "null terminator expected in _Nt_checked array initializer">;
+
+  def err_initializer_not_null_terminated_for_nt_checked_string_literal : Error<
+     "initializer-string for _Nt_checked char array is too long">;
```
Comment (dtarditi) [inline include/clang/Basic/DiagnosticSemaKinds.td]: I think this could be a little confusing to a developer.   I would suggest "initializer string for _Nt_checked char array does not initialize last array element to null."

File: lib/Sema/SemaDecl.cpp [hunk 7]
```diff
@@ -11065,6 +11065,16 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit,
       getCurFunction()->markSafeWeakUse(Init);
   }
 
+  // Checked C: All initializers to _NT_CHECKED type arrays must be null
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: Please replace `_NT_CHECKED` with `_Nt_checked`.

File: lib/Sema/SemaDecl.cpp [hunk 8]
```diff
@@ -11262,6 +11272,111 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit,
   CheckCompleteVariableDeclaration(VDecl);
 }
 
+/// Checked C: Validate that the _Nt_checked type array initializers must be
+/// null terminated VDeclType - CanonicalType of the declared variable that is
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: Please put the argument descriptions on separate lines.

File: lib/Sema/SemaDecl.cpp [hunk 9]
```diff
@@ -11262,6 +11272,111 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit,
...
+/// initialized Init - initializer expression for a declaration with VDeclType
+bool Sema::ValidateNTCheckedType(ASTContext &Ctx, QualType VDeclType,
+                                 Expr *Init) {
+  if (!Init) {
+    return true;
+  }
+  const Type *current = VDeclType.getTypePtr();
+  switch (current->getTypeClass()) {
+  case Type::Pointer: {
+    // Pointers to _Nt_checked types (nt_arry_ptr) are enforced to point to
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: Typo in `_nt_array_ptr`.

Unnecessary space between `null-` and termianted.

File: include/clang/Basic/DiagnosticSemaKinds.td (hunk 6 above)
Comment (Prabhuk) [reply include/clang/Basic/DiagnosticSemaKinds.td]: David, this happens only if the string literal assigned to the character array is longer than the size specified in the declaration. The warning shown by clang is "initializer-string for char array is too long". I tried to stick to a similarly worded message for checked C. Please let me know if you still want the change you've suggested here.

**Comment by Prabhuk** [comment]
All the suggested changes from previous review are completed except for the error message changes. Please see my comment for leaving the error message as it is.
